### PR TITLE
feat: inactivity tracking pallet (chain inactivity tracking)

### PR DESCRIPTION
**Pull request by chexware**
This PR adds chain inactivity tracking to `pallet-inactivity-tracking`

TO DO:
- [x] Fix and refactor unit tests
- [x] Add manual disabling container chains from block production RPC call to Dancelight  in dev mode
- [x] Prevent collators assigned to parathreads to be marked as inactive
- [x] Add unit tests covering parathread-related logic
- [x] Add integration tests covering parathread-related edge cases

### CODE REVIEW

**Comment by girazoki** [comment]
As I said internally let's not include both chain and collatot inactivy in the same PR

File: pallets/inactivity-tracking/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            // Process the orchestrator chain block author (if it exists)
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight += Self::on_author_noted(orchestrator_chain_author);
+            }
+
+            // Update inactive collator records only after a session has ended
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
```
Comment (girazoki) [inline pallets/inactivity-tracking/src/lib.rs:133]: A way to avoid 2 reads every block is doing this inside `apply_own_session`. However I dont dislike this

Comment (girazoki) [reply pallets/inactivity-tracking/src/lib.rs:133]: or inside the function `before_session_ending`

File: pallets/inactivity-tracking/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn process_ended_session(
+            unprocessed_session_id: SessionIndex,
+            current_session_id: SessionIndex,
+        ) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(0, 4);
+            ActiveCollators::<T>::insert(
+                unprocessed_session_id,
```
Comment (girazoki) [inline pallets/inactivity-tracking/src/lib.rs:154]: this is wrong since you are here passing unprocessed_session_id, but the active collators have already changed. I am assuming that what you want to do is `current_session_id`

**Comment by chexware** [comment]
> As I said internally let's not include both chain and collator inactivity in the same PR

@girazoki  Its updated now! Collators inactivity tracking is in https://github.com/moondance-labs/tanssi/pull/922 and this branch merges into it

File: pallets/inactivity-tracking/src/lib.rs (hunk 1 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: Good point...  Using `before_session_ending ` should remove the need for storing `LastUnprocessedSession` session too... I will add the `SessionHandler` hook for dealing with  chain inactivity first and update this afterwards

File: pallets/inactivity-tracking/src/lib.rs (hunk 2 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs:181]: No worries! Will be resolving this conversation and  moving all related changes to https://github.com/moondance-labs/tanssi/tree/aleks-collator-inactivity-tracking

File: pallets/inactivity-tracking/src/lib.rs (hunk 1 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: Changes will be applied in https://github.com/moondance-labs/tanssi/tree/aleks-collator-inactivity-tracking

File: chains/orchestrator-relays/runtime/dancelight/src/tests/inactivity_tracking.rs [hunk 3]
```diff
@@ -81,6 +83,19 @@ fn get_collators_set(
...
 
+fn get_chains_set(
+    chains: Vec<ParaId>,
+) -> BoundedBTreeSet<ParaId, <Runtime as pallet_inactivity_tracking::Config>::MaxContainerChains> {
+    let mut chains_set: BoundedBTreeSet<
+        ParaId,
+        <Runtime as pallet_inactivity_tracking::Config>::MaxContainerChains,
+    > = BoundedBTreeSet::new();
+    chains.iter().for_each(|collator| {
+        chains_set.try_insert(*collator).ok();
```
Comment (tmpolaczyk) [inline chains/orchestrator-relays/runtime/dancelight/src/tests/inactivity_tracking.rs:94]: Since this is for tests, panic on error.

```rust
for para_id in chains {
    chains_set.try_insert(para_id).unwrap()
}
```

Comment (chexware) [reply chains/orchestrator-relays/runtime/dancelight/src/tests/inactivity_tracking.rs]: Sure updated in https://github.com/moondance-labs/tanssi/pull/899/commits/ec3d31e2ccfed1f1b4406b768d22b0cf94b2c131

File: pallets/inactivity-tracking/src/lib.rs [hunk 4]
```diff
@@ -304,6 +315,40 @@ pub mod pallet {
...
+                        ForSession::Current,
+                    )
+                    .iter()
+                    .for_each(|(para_id, collator_ids)| {
+                        if !active_chains.contains(para_id) {
+                            collator_ids.iter().for_each(|collator_id| {
+                                if !active_collators.contains(collator_id) {
+                                    let _ = active_collators
+                                        .try_insert(collator_id.clone())
+                                        .map_err(|_| Error::<T>::MaxCollatorsPerSessionReached);
```
Comment (girazoki) [inline pallets/inactivity-tracking/src/lib.rs:342]: this ensures that  if the chain is inactive, we mark its collators as active?

Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs:342]: Yes!

File: pallets/inactivity-tracking/src/lib.rs [hunk 5]
```diff
@@ -304,6 +315,40 @@ pub mod pallet {
...
+                }
+            }
+            let active_chains = <ActiveContainerChainsForCurrentSession<T>>::get();
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    T::CurrentCollatorsFetcher::container_chains_with_collators(
+                        ForSession::Current,
+                    )
+                    .iter()
+                    .for_each(|(para_id, collator_ids)| {
```
Comment (tmpolaczyk) [inline pallets/inactivity-tracking/src/lib.rs:336]: Just use a for loop, then you can use `?` to return on errors

File: pallets/inactivity-tracking/src/lib.rs [hunk 6]
```diff
@@ -154,6 +156,11 @@ pub mod pallet {
     pub type ActiveCollatorsForCurrentSession<T: Config> =
         StorageValue<_, BoundedBTreeSet<T::CollatorId, T::MaxCollatorsPerSession>, ValueQuery>;
 
+    /// A list of active container chains for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveContainerChainsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedBTreeSet<ParaId, T::MaxContainerChains>, ValueQuery>;
```
Comment (tmpolaczyk) [inline pallets/inactivity-tracking/src/lib.rs:162]: So this starts empty, then using `on_chain_noted` active chains are inserted here, and then when the session ends it is "processed" and then cleared again.

The processing filters inactive collators. If a collator was assigned to an inactive chain it is not considered inactive.

When we add parathreads it is possible for one collator to be assigned to more than one chain, what happens in that case? If one chain is inactive but the other chain not? Or maybe this mark inactive feature doesn't make sense with parathreads.

File: pallets/inactivity-tracking/src/lib.rs (hunk 5 above)
Comment (tmpolaczyk) [reply pallets/inactivity-tracking/src/lib.rs:336]: Didn't verify but maybe this works and is cleaner:

```rust
        // Inactive collator assigned to inactive chain stays inactive
        if active_chains.contains(&para_id) {
            continue;
        }

        // Inactive collator assigned to inactive chain becomes active
        for collator_id in collator_ids {
            if !active_collators.contains(&collator_id) {
                active_collators
                    .try_insert(collator_id.clone())
                    .map_err(|_| Error::<T>::MaxCollatorsPerSessionReached)?;
            }
        }
```

File: pallets/inactivity-tracking/src/lib.rs [hunk 7]
```diff
@@ -327,11 +372,35 @@ pub mod pallet {
             );
             total_weight
         }
+
+        /// Internal update the current session active chains records.
+        /// This function is called when a container chain is noted.
+        pub fn on_chain_noted(chain_id: ParaId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveContainerChainsForCurrentSession<T>>::try_mutate(
```
[the remark this answers is not in the document]
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs:388]: Updated the code to disable inactivity tracking if `ActiveContainerChainsForCurrentSession` overflows in https://github.com/moondance-labs/tanssi/pull/899/commits/de2e097984dd1136169c022c8657b0667008df5a and added unit test for it in https://github.com/moondance-labs/tanssi/pull/899/commits/22b72ab61cfd6a647a074b38e6edd47e958f93e5

File: pallets/inactivity-tracking/src/lib.rs (hunk 4 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: Updated the code to disable inactivity tracking if `ActiveCollatorsForCurrentSession` while processing inactive chains and added test for it in https://github.com/moondance-labs/tanssi/pull/899/commits/1ea2a0b3266a7ea66d31fd1d84a02ff3e9b92644 and https://github.com/moondance-labs/tanssi/pull/899/commits/428c36c8d4f8afeb8d2c9c53b341d3efb86ce357

File: pallets/inactivity-tracking/src/lib.rs (hunk 6 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs:162]: Very good point! If there is a possibility for a collator to produce blocks on multiple chain perhaps we can change how `process_inactive_chains_for_current_session`  work.
1) Iterate over all chains an create 2 vectors of collators - one for active chains and one for inactive.
2) Iterate over all inactive collators and add them to `ActiveCollatorsForCurrentSession` only if they 
	- aren't already there
	- aren't in the active chains collators vector
 This should solve the issue

Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs:166]: Added additional check that will prevent parathreads from being processed as inactive parachains for now in https://github.com/moondance-labs/tanssi/pull/899/commits/a3869cb10695bf6dc60214f54ea0bb63776fbf94 .

File: chains/orchestrator-paras/runtime/dancebox/src/tests/inactivity_tracking.rs [hunk 8]
```diff
@@ -66,6 +68,19 @@ fn get_collators_set(
     collator_set
 }
 
+fn get_chains_set(
```
Comment (girazoki) [inline chains/orchestrator-paras/runtime/dancebox/src/tests/inactivity_tracking.rs:71]: Same thing as above

File: chains/orchestrator-relays/node/tanssi-relay-service/src/dev_service.rs [hunk 9]
```diff
@@ -341,8 +364,32 @@ where
...
+            .expect("Should be able to query aux storage; qed")
+            .unwrap_or(Vec::<ParaId>::new().encode());
+        let mock_excluded_container_chains: Vec<ParaId> =
+            Decode::decode(&mut new_excluded_container_chains_value.as_slice())
+                .expect("Vector non-decodable");
+
         // iterate over every core|para pair
         for (core, para) in claim_queue {
+            let mut para = para.clone();
+            para.retain(|x| !mock_excluded_container_chains.contains(x));
```
Comment (girazoki) [inline chains/orchestrator-relays/node/tanssi-relay-service/src/dev_service.rs:392]: nice, comment please

File: pallets/inactivity-tracking/src/lib.rs [hunk 10]
```diff
@@ -304,6 +319,53 @@ pub mod pallet {
...
+            let active_chains = <ActiveContainerChainsForCurrentSession<T>>::get();
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    let container_chains_with_collators =
+                        T::CurrentCollatorsFetcher::container_chains_with_collators(
+                            ForSession::Current,
+                        );
+                    for (para_id, collator_ids) in container_chains_with_collators.iter() {
+                        if !active_chains.contains(para_id)
+                            && !T::ParathreadHelper::is_parathread(para_id)
```
Comment (girazoki) [inline pallets/inactivity-tracking/src/lib.rs:342]:
```suggestion
                            && !T::ParaFilter::is_parathread(para_id)
```
let's be more generic, and call it filter

File: pallets/inactivity-tracking/src/tests.rs [hunk 11]
```diff
@@ -562,3 +614,273 @@ fn inactivity_tracking_is_disabled_if_current_active_collators_storage_overflows
         );
     });
 }
+
+#[test]
+fn inactivity_tracking_is_disabled_if_ctive_chains_storage_overflows() {
```
Comment (girazoki) [inline pallets/inactivity-tracking/src/tests.rs:619]:
```suggestion
fn inactivity_tracking_is_disabled_if_active_chains_storage_overflows() {
```

File: pallets/inactivity-tracking/src/tests.rs [hunk 12]
```diff
@@ -562,3 +614,273 @@ fn inactivity_tracking_is_disabled_if_current_active_collators_storage_overflows
...
+        assert_eq!(
+            ActiveContainerChainsForCurrentSession::<Test>::get().is_empty(),
+            true
+        );
+        assert_eq!(
+            ActiveCollatorsForCurrentSession::<Test>::get().is_empty(),
+            true
+        );
+        roll_to(SESSION_BLOCK_LENGTH);
+        // Since we have one container chain with id CONTAINER_CHAIN_ID_3 which is a parathread
```
Comment (girazoki) [inline pallets/inactivity-tracking/src/tests.rs:762]: Wait should not this be the opposite? i.e.:

- if assigned to a parathread, we dont care if a collator is active or inactive
- if assigned to a parachain, we acre to see if a collator is active or inactive

File: pallets/inactivity-tracking/src/lib.rs (hunk 10 above)
Comment (girazoki) [reply pallets/inactivity-tracking/src/lib.rs:342]: so basically I would make it inject a vec of para ids and return a vec of paraids

File: pallets/inactivity-tracking/src/tests.rs (hunk 12 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/tests.rs:762]: @girazoki with the current setup we only use the filter when tracking chain inactivity (so for parathreads we ignore inactivty tracking). Do we want to ignore all parathread activity?

File: pallets/inactivity-tracking/src/lib.rs (hunk 10 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: Naming updated in https://github.com/moondance-labs/tanssi/pull/899/commits/caa56e8e3a9a064a6ad44f0989b3b7423369eb6f

File: chains/orchestrator-paras/runtime/dancebox/src/lib.rs [hunk 13]
```diff
@@ -1708,14 +1712,27 @@ impl pallet_multisig::Config for Runtime {
     type WeightInfo = weights::pallet_multisig::SubstrateWeight<Runtime>;
 }
 
+pub struct DanceboxParathreadHelper;
```
[the remark this answers is not in the document]
Comment (chexware) [reply chains/orchestrator-paras/runtime/dancebox/src/lib.rs]: Refactored in https://github.com/moondance-labs/tanssi/pull/899/commits/532e81b8fbf93cdab325134a78cfee10136eaa6e

File: pallets/inactivity-tracking/src/tests.rs (hunk 12 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/tests.rs]: Updated the code and added a explanation in https://github.com/moondance-labs/tanssi/pull/899/commits/85a78a344f8b3e1ba3574fa30c1e6eb23a95cde8

File: primitives/parathread-filter-common/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+impl<Runtime: pallet_session::Config + pallet_registrar::Config> ParathreadHelper
+    for ExcludeAllParathreadsFilter<Runtime>
+{
+    fn is_parathread(para_id: &ParaId) -> bool {
+        pallet_registrar::Pallet::<Runtime>::session_container_chains(
+            pallet_session::Pallet::<Runtime>::current_index().into(),
+        )
+        .parathreads
+        .iter()
+        .any(|(id, _)| id == para_id)
```
Comment (tmpolaczyk) [inline primitives/parathread-filter-common/src/lib.rs:34]: I believe @girazoki already mentioned it, but its better for this function to take a `&mut Vec<ParaId>` and filter all at once. And actually it may be easier to just return the list of parathreads and do the filtering in the pallet.

Comment (chexware) [reply primitives/parathread-filter-common/src/lib.rs]: Refactored in https://github.com/moondance-labs/tanssi/pull/899/commits/14c88e43259860951ca9797cc60e1e845dbd18a2

File: pallets/inactivity-tracking/src/lib.rs (hunk 10 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: ParaFilter refactored to return a vector of parathreads in https://github.com/moondance-labs/tanssi/pull/899/commits/14c88e43259860951ca9797cc60e1e845dbd18a2

File: pallets/inactivity-tracking/src/lib.rs [hunk 15]
```diff
@@ -331,19 +331,26 @@ pub mod pallet {
...
+            // Fetching the parathreads for the current session and remove them
+            // from the active chains array. In this way we handle all parathreads
+            // as inactive chains. This solution would only work if a collator either:
+            // - is assigned to one chain only
+            // - is assigned to multiple chains but all of them are parathreads
+            let parathreads_for_session = T::ParaFilter::get_parathreads_for_session();
+            active_chains
+                .clone()
+                .into_inner()
+                .retain(|x| !parathreads_for_session.contains(x));
```
Comment (tmpolaczyk) [inline pallets/inactivity-tracking/src/lib.rs:343]: This doesn't do anything because of the clone, I hope tests fail

Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: My bad updated in https://github.com/moondance-labs/tanssi/pull/899/commits/c06a56bf979f032c62401621a7641e1360b6c97b

Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: Also additional test that will fail if parathread collators are processed as inactive in https://github.com/moondance-labs/tanssi/pull/899/commits/99a6a37b625ba4eb69720cbb0b19eb8e054c66bd

File: pallets/inactivity-tracking/src/lib.rs [hunk 16]
```diff
@@ -304,6 +319,60 @@ pub mod pallet {
...
+                }
+            }
+            let mut active_chains = <ActiveContainerChainsForCurrentSession<T>>::get().into_inner();
+            // Fetching the parathreads for the current session and remove them
+            // from the active chains array. In this way we handle all parathreads
+            // as inactive chains. This solution would only work if a collator either:
+            // - is assigned to one chain only
+            // - is assigned to multiple chains but all of them are parathreads
+            let parathreads_for_session = T::ParaFilter::get_parathreads_for_session();
+            active_chains.retain(|x| !parathreads_for_session.contains(x));
```
Comment (evgeny-s) [inline pallets/inactivity-tracking/src/lib.rs:340]: This probably can be optimized a bit in terms of performance, since we do loop + contains for vec (another loop), it is O(m x n) time complexity. 
If we have at least 1000 container chains and at least 100 active chains it will lead to 100k iterations. Not sure if those are the realistic numbers, but I think we can convert the vec to hashset before do contains.

File: pallets/inactivity-tracking/src/lib.rs [hunk 17]
```diff
@@ -304,6 +319,60 @@ pub mod pallet {
...
+                        T::CurrentCollatorsFetcher::container_chains_with_collators(
+                            ForSession::Current,
+                        );
+
+                    for (para_id, collator_ids) in container_chains_with_collators.iter() {
+                        if !active_chains.contains(para_id) {
+                            // Collators assigned to inactive chain are added
+                            // to the current active collators storage
+                            for collator_id in collator_ids {
+                                if !active_collators.contains(&collator_id) {
```
Comment (evgeny-s) [inline pallets/inactivity-tracking/src/lib.rs:354]: Do we really need this `contains` here?
`active_collators.try_insert` already should raise a error in case of either the limit is reached or collator_id in the set.

File: pallets/inactivity-tracking/src/lib.rs (hunk 16 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: Very good point! Updated `get_parathreads_for_session` to return `BTreeSet` in https://github.com/moondance-labs/tanssi/pull/899/commits/7cee1f04a04ead2e49c105029cbe691858f8e26b

File: pallets/inactivity-tracking/src/lib.rs (hunk 17 above)
Comment (chexware) [reply pallets/inactivity-tracking/src/lib.rs]: True, don't think we need it. Removed it in https://github.com/moondance-labs/tanssi/pull/899/commits/f36a655b77148031f869899078a1c4c1b4ed91f3

File: chains/orchestrator-relays/runtime/dancelight/src/lib.rs [hunk 18]
```diff
@@ -1882,10 +1886,12 @@ impl pallet_inactivity_tracking::Config for Runtime {
     type RuntimeEvent = RuntimeEvent;
     type CollatorId = AccountId;
     type MaxInactiveSessions = ConstU32<5>;
```
[the remark this answers is not in the document]
Comment (chexware) [reply chains/orchestrator-relays/runtime/dancelight/src/lib.rs]: Updated in https://github.com/moondance-labs/tanssi/pull/899/commits/7d8f2cecfcadc354ca1b7e93eee8fb44499f7665
